### PR TITLE
Always pass the correct batch index to the automatic optimization loop

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes https://twitter.com/MF_FOOM/status/1705265642788749773

----
:books: Documentation preview :books:: https://pytorch-lightning--18619.org.readthedocs.build/en/18619/

cc @borda @carmocca @justusschock

### CODE REVIEW

File: src/lightning/pytorch/trainer/configuration_validator.py [hunk 1]
```diff
@@ -152,16 +153,19 @@ def __verify_manual_optimization_support(trainer: "pl.Trainer", model: "pl.Light
...
-    """Check if the current `training_step` is requesting `dataloader_iter`."""
-    if is_param_in_hook_signature(model.training_step, "dataloader_iter", explicit=True):
-        for hook in ("on_train_batch_start", "on_train_batch_end"):
-            if is_overridden(hook, model):
-                rank_zero_warn(
-                    f"The `batch_idx` argument in `{type(model).__name__}.{hook}` hook may"
-                    " not match with the actual batch index when using a `dataloader_iter`"
-                    " argument in your `training_step`."
-                )
+def __warn_dataloader_iter_limitations(model: "pl.LightningModule") -> None:
```
Comment (carmocca) [inline src/lightning/pytorch/trainer/configuration_validator.py:156]: This change is actually unrelated to the bugfix, but the bugfix made me think that `dataloader_iter` will also be broken with gradient accumulation.

Unfortunately, there's no way for us to know whether the user will fetch more than one batch per step. Given how tricky and experimental this feature is, I would say it's good to be this vocal about the limitations.

File: src/lightning/pytorch/trainer/configuration_validator.py [hunk 2]
```diff
@@ -152,16 +153,20 @@ def __verify_manual_optimization_support(trainer: "pl.Trainer", model: "pl.Light
...
+    """Check if `dataloader_iter is enabled`."""
+    if any(
+        is_param_in_hook_signature(step_fn, "dataloader_iter", explicit=True)
+        for step_fn in (model.training_step, model.validation_step, model.predict_step, model.test_step)
+        if step_fn is not None
+    ):
+        rank_zero_warn(
+            "You are using the `dataloader_iter` step flavor. If you consume the iterator more than once per step, the"
```
Comment (awaelchli) [inline src/lightning/pytorch/trainer/configuration_validator.py:164]: Shouldn't this warning suggest that you remove the batch_idx argument if you want to avoid the warning?

File: src/lightning/pytorch/trainer/configuration_validator.py [hunk 3]
```diff
@@ -152,16 +153,20 @@ def __verify_manual_optimization_support(trainer: "pl.Trainer", model: "pl.Light
...
+            " `batch_idx` argument in any hook that takes it will not match with the batch index of the last batch"
+            " consumed. This might have unforeseen effects on callbacks or code that expects to get the correct index."
+            " This will also no work well with gradient accumulation. This feature is very experimental and subject to"
+            " change. Here be dragons."
```
Comment (awaelchli) [inline src/lightning/pytorch/trainer/configuration_validator.py:168]: This could be a PossibleUserWarning

**Review by awaelchli — APPROVED** [review]

File: src/lightning/pytorch/trainer/configuration_validator.py (hunk 2 above)
Comment (carmocca) [reply src/lightning/pytorch/trainer/configuration_validator.py:164]: No, it will still be a problem for all the hooks in the LM, LDM, or callbacks that also take the `batch_idx`. And since the user might not own their implementation it is a generally unavoidable issue

**Review by Borda — APPROVED** [review]
